packages/jazz-tools/src/schema/coMap.ts

```diff
@@ -0,0 +1,364 @@
+> | null>
+  ? NonNullable<V>
+  : V extends co<CoMapSchema | undefined>
+    ? CoMapInstance<NonNullable<UnCo<V>>> | null
+    : V;
+
+export type CoMapInit<Map extends object> = {
+  [Key in CoKeys<Map> as undefined extends Map[Key]
+    ? never
+    : IfCo<Map[Key], Key>]: ForceRequiredRef<Map[Key]>;
+} & {
+  [Key in CoKeys<Map> as IfCo<Map[Key], Key>]?: ForceRequiredRef<Map[Key]>;
+};
+
+export class CoMapInstance<M extends CoMapSchema> extends CoValuePrototype {
+  declare $raw: RawCoMap;
+  declare $schema: M;
+  declare $id: ID<M>;
+  declare $resolutionNode: CoValueResolutionNode<M> | undefined;
+  refs = new Map<string, CoMapInstance<any>>();
+  private $lastUpdateTx: number;
+
+  constructor(
+    schema: M,
+    raw: RawCoMap,
+    resolutionNode?: CoValueResolutionNode<M>,
+  ) {
+    super();
+    this.$schema = schema;
+    this.$raw = raw;
+    this.$lastUpdateTx = raw.totalProcessedTransactions;
+    this.$id = raw.id as unknown as ID<M>;
+    this.$resolutionNode = resolutionNode;
+  }
+
+  static fromInit<M extends CoMapSchema>(
+    schema: M,
+    init: Simplify<CoMapInit<M>> | undefined,
+    owner: Account | Group,
+    uniqueness?: CoValueUniqueness,
+  ) {
+    const { raw, refs } = createCoMapFromInit(init, owner, schema, uniqueness);
+
+    return CoMapInstance.fromRaw(schema, raw, refs);
+  }
+
+  static fromRaw<M extends CoMapSchema>(
+    schema: M,
+    raw: RawCoMap,
+    refs?: Map<string, CoMapInstance<any> | undefined>,
+    resolutionNode?: CoValueResolutionNode<M>,
+  ) {
+    const instance = Object.create(
+      new CoMapInstance(schema, raw, resolutionNode),
+    );
+
+    const isRecord = schema.getFieldDescriptor(ItemsSym) !== undefined;
+    const fields = isRecord ? raw.keys() : Object.keys(schema);
+
+    for (const key of fields) {
+      Object.defineProperty(instance, key, {
+        value: getValue(raw, schema, key),
+        writable: false,
+        enumerable: true,
+        configurable: true,
+      });
+    }
+
+    if (refs) {
+      for (const [key, value] of refs.entries()) {
+        if (value) {
+          instance._fillRef(key as any, value);
+        }
+      }
+    }
+
+    return instance as CoMapInstance<M> & Simplify<CoMapInit<M>>;
+  }
+
+  _fillRef<K extends CoKeys<M>>(key: K, value: CoMapInstance<any>) {
+    const descriptor = this.$schema.getFieldDescriptor(key);
+
+    if (descriptor && isRefEncoded(descriptor)) {
+      this.refs.set(key, value);
+      Object.defineProperty(this, key, {
+        value,
+        writable: false,
+        enumerable: true,
+        configurable: true,
+      });
+    } else {
+      throw new Error(`Field ${key} is not a reference`);
+    }
+  }
+
+  $set<K extends CoKeys<M>>(key: K, value: M[K]) {
+    setValue(this.$raw, this.$schema, key, value as JsonValue);
+  }
+
+  $updated(refs?: Map<string, CoMapInstance<any> | undefined>) {
+    if (this.$lastUpdateTx === this.$raw.totalProcessedTransactions && !refs) {
+      return this;
+    }
```

Comment:
This is one of the key APIs.

It preserves the current instance unless:
- Some new transactions are coming from the core value
- new refs are provided